contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,295 @@
+
+
+### Proposed Core Metrics API:
+
+An important difference between the current summary api and the proposed core metrics api is that per-pod stats in the core metrics api contain only usage data, and not capacity-related statistics.  This is more accurate since a pod's resource capacity is really defined by its "requests" and "limits", and it is a better reflection of how components use these metrics.  The kubelet, for example, finds which resources are constrained using node-level capacity and availability data, and then chooses which pods to take action on based on the pod's usage of the constrained resource.  If neccessary, capacity for resources a pod consumes can still be correlated with node-level resources using this format of stats.
+
+```go
+// CoreStats is a top-level container for holding NodeStats and PodStats.  
+type CoreStats struct {  
+  // Overall node resource stats.  
+  Node NodeResources `json:"node"`  
+  // Per-pod usage stats.  
+  Pods []PodUsage `json:"pods"`  
+}  
+
+// NodeStats holds node-level stats.  NodeStats contains capacity and availibility for Node Resources.  
+type NodeResources struct {  
+  // The filesystem device used by node k8s components.  
+  // +optional  
+  KubeletFsDevice string `json:"kubeletfs,omitempty"`  
+  // The filesystem device used by node runtime components.  
+  // +optional  
+  RuntimeFsDevice string `json:"runtimefs,omitempty"`  
+  // Stats pertaining to cpu resources.  
+  // +optional  
+  CPU *CpuResources `json:"cpu,omitempty"`  
+  // Stats pertaining to memory (RAM) resources.  
+  // +optional  
+  Memory *MemoryResources `json:"memory,omitempty"`  
+  // Stats pertaining to node filesystem resources.  
+  // +optional  
+  Filesystems []FilesystemResources `json:"filesystems, omitempty" patchStrategy:"merge" patchMergeKey:"device"`  
+}  
+
+// CpuResources containes data about cpu resource usage  
+type CpuResources struct {  
+  // The number of cores in this machine.  
+  NumCores int `json:"numcores"`  
+  // The current Usage of CPU resources  
+  Usage *CpuUsage `json:"cpuusage,omitempty"`  
+}  
+
+// MemoryResources contains data about memory resource usage.  
+type MemoryResources struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // The memory capacity, in bytes  
+  CapacityBytes *uint64 `json:"capacitybytes,omitempty"`  
+  // The available memory, in bytes  
+  // This is the number of bytes which are not included in the working set memory
+  // Working set memory includes recently accessed memory, dirty memory, and kernel memory.
+  AvailableBytes *uint64 `json:"availablebytes,omitempty"`  
+}  
+
+// FilesystemResources contains data about filesystem disk resources.  
+type FilesystemResources struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // The device that this filesystem is on  
+  Device string `json:"device"`  
```

Comment:
nit: what is the spec for this string? (i.e. relative to root? relative to /dev/?)  How is this used?